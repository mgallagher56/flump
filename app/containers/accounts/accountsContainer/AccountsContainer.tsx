import type { FC } from 'react';
import { Fragment, useMemo } from 'react';

import { useLoaderData } from '@remix-run/react';
import { useTranslation } from 'react-i18next';
import AccountsCard from '~/components/core/cards/AccountsCard';
import FLPBox from '~/components/core/structure/FLPBox';
import FLPHeading from '~/components/core/typography/FLPHeading';
import type { loader } from '~/routes/app.accounts._index';

import AddEditAccountsDialogBtn from '../../../components/dialogs/addEditAccountsDialog.tsx/AddEditAccountsDialog';
import { AccountTypeEnum, isAccountTypeValid } from '../utils';

const { CURRENT, SAVING, MORTGAGE, CREDIT_CARD, LOAN, OWED } = AccountTypeEnum;

const AccountsContainer: FC = () => {
  const { t } = useTranslation();
  const { accounts } = useLoaderData<typeof loader>();
  const accountTypeArr = useMemo(() => [CURRENT, SAVING, CREDIT_CARD, MORTGAGE, LOAN, OWED], []);

  return (
    <>
      <FLPBox mb={5} display="flex" flexDirection="row" justifyContent="space-between">
        <FLPHeading size="xl" as="h1">
          {t('accounts')}
        </FLPHeading>
        <AddEditAccountsDialogBtn />
      </FLPBox>
      {accountTypeArr.map((accountType) => {
        return (
          <Fragment key={accountType}>
            {isAccountTypeValid(accountType, accounts) && (
              <FLPHeading as="h2" size="lg">
                {accountType}
              </FLPHeading>
            )}
            <FLPBox display="flex" flexWrap="wrap" gap={5}>
              {accounts.map(
                ({ name, id, type }: { name?: string; id?: string; type?: AccountTypeEnum | string }) =>
                  type === accountType && <AccountsCard key={id} accountId={id} name={name} type={type} />
              )}
            </FLPBox>
          </Fragment>
        );
      })}
    </>
  );
};

export default AccountsContainer;
