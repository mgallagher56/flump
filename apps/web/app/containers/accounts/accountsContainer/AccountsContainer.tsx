import type { FC } from 'react';
import { Fragment, useMemo } from 'react';

import { useTranslation } from 'react-i18next';
import { useLoaderData } from 'react-router';
import AccountsCard from '~/components/core/cards/AccountsCard';
import FLPBox from '~/components/core/structure/FLPBox';
import FLPHeading from '~/components/core/typography/FLPHeading';
import AddEditAccountsDialogBtn from '~/components/dialogs/addEditAccountsDialog.tsx/AddEditAccountsDialog';
import { AccountTypeEnum, isAccountTypeValid } from '~/containers/accounts/utils';
import type { loader } from '~/routes/app.accounts._index';

const { CURRENT, SAVING, MORTGAGE, CREDIT_CARD, LOAN, OWED } = AccountTypeEnum;

const AccountsContainer: FC = () => {
  const { t } = useTranslation();
  const { accounts } = useLoaderData<typeof loader>();
  const accountTypeArr = useMemo(() => [CURRENT, SAVING, CREDIT_CARD, MORTGAGE, LOAN, OWED], []);

  return (
    <>
      <FLPBox display="flex" flexDirection="row" justifyContent="space-between" my={5}>
        <FLPHeading as="h1" size="xl">
          {t('accounts')}
        </FLPHeading>
        <AddEditAccountsDialogBtn />
      </FLPBox>
      <FLPBox display="flex" flexWrap="wrap" gap={10}>
        {accountTypeArr.map((accountType) => {
          return (
            isAccountTypeValid(accountType, accounts) && (
              <Fragment key={accountType}>
                <FLPBox>
                  <FLPHeading as="h2" size="lg">
                    {accountType}
                  </FLPHeading>
                  <FLPBox display="flex" flexWrap="wrap" gap={5}>
                    {accounts.map(
                      ({ name, id, type }: { name?: string; id?: string; type?: AccountTypeEnum | string }) =>
                        type === accountType && <AccountsCard key={id} accountId={id} name={name} type={type} />
                    )}
                  </FLPBox>
                </FLPBox>
              </Fragment>
            )
          );
        })}
      </FLPBox>
    </>
  );
};

export default AccountsContainer;
