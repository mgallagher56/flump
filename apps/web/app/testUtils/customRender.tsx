// create a custom render function with ChakraProvider
import { type ReactNode } from 'react';

import { type RenderOptions, type RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { Provider } from '~/components/ui/provider';

const user = userEvent.setup();

const customRender = (ui: ReactNode, options?: RenderOptions): RenderResult & { user: UserEvent } => {
  const wrappedUi = <Provider>{ui}</Provider>;

  return {
    ...rtlRender(wrappedUi, options),
    user
  };
};

export default customRender;
