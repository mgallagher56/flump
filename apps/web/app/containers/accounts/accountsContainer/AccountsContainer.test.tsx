import type { ReactNode } from 'react';

import { fireEvent, waitFor } from '@testing-library/react';
import { vi } from 'vitest';
import customRender from '~/testUtils/customRender';

import mockAccounts from '__mocks__/accounts';

import AccountsContainer from './AccountsContainer';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn(),
  mockUseRevalidator: vi.fn(() => ({ revalidate: vi.fn() }))
}));

vi.mock('react-router', async () => {
  const actual: Record<string, unknown> = await vi.importActual('react-router');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData,
    useRevalidator: mocks.mockUseRevalidator,
    Form: ({ children }: { children: ReactNode }) => <form>{children}</form>,
    useSubmit: () => ({ onSubmit: vi.fn() })
  };
});

vi.mock('app/components/core/cards/AccountsCard', () => ({ default: () => 'AccountsCard' }));
vi.mock('app/components/core/input/FLPInput', () => ({ default: () => 'FLPInput' }));

describe('<AccountsContainer />', () => {
  test('should render as expected when no accounts exist', () => {
    mocks.mockUseLoaderData.mockReturnValue({ accounts: [] });
    const { container } = customRender(<AccountsContainer />);
    expect(container).toMatchSnapshot();
  });
 
  test('should render as expected when accounts exist', () => {
    mocks.mockUseLoaderData.mockReturnValue({ accounts: mockAccounts });
    const { container } = customRender(<AccountsContainer />);
    expect(container).toMatchSnapshot();
  });

  test('should call supabase functions to add an account', async () => {
    const { getAllByText } = customRender(<AccountsContainer />);
    const addAccountModalBtn = getAllByText('addAccount')[0];
    expect(addAccountModalBtn).toBeDefined();
    fireEvent.click(addAccountModalBtn);
    await waitFor(() => {
      expect(getAllByText('addAccount')[0]).toBeDefined();
    });
  });
});
