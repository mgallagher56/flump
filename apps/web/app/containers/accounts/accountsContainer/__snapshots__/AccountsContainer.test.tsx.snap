// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccountsContainer /> > should render as expected when accounts exist 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;
    {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
    {let k=y==="class",S=k&&m?d.map(f=&gt;
    m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-u7ajzs"
  >
    <h1
      class="chakra-heading css-mytsjf"
    >
      accounts
    </h1>
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="chakra-dialog__trigger css-0"
      data-part="trigger"
      data-scope="dialog"
      data-state="closed"
      dir="ltr"
      id="dialog::r1::trigger"
      type="button"
    >
      <button
        class="chakra-button css-teolye"
        type="button"
      >
        addAccount
      </button>
    </button>
  </div>
  <div
    class="css-1jswdmc"
  >
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-e8b6d6"
      >
        Current
      </h2>
      <div
        class="css-15fx9a5"
      >
        AccountsCard
        AccountsCard
      </div>
    </div>
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-e8b6d6"
      >
        Saving
      </h2>
      <div
        class="css-15fx9a5"
      >
        AccountsCard
        AccountsCard
        AccountsCard
      </div>
    </div>
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-e8b6d6"
      >
        Mortgage
      </h2>
      <div
        class="css-15fx9a5"
      >
        AccountsCard
      </div>
    </div>
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-e8b6d6"
      >
        Loan
      </h2>
      <div
        class="css-15fx9a5"
      >
        AccountsCard
      </div>
    </div>
    <div
      class="css-0"
    >
      <h2
        class="chakra-heading css-e8b6d6"
      >
        Owed
      </h2>
      <div
        class="css-15fx9a5"
      >
        AccountsCard
      </div>
    </div>
  </div>
</div>
`;

exports[`<AccountsContainer /> > should render as expected when no accounts exist 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;
    {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
    {let k=y==="class",S=k&&m?d.map(f=&gt;
    m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-u7ajzs"
  >
    <h1
      class="chakra-heading css-mytsjf"
    >
      accounts
    </h1>
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="chakra-dialog__trigger css-0"
      data-part="trigger"
      data-scope="dialog"
      data-state="closed"
      dir="ltr"
      id="dialog::r0::trigger"
      type="button"
    >
      <button
        class="chakra-button css-teolye"
        type="button"
      >
        addAccount
      </button>
    </button>
  </div>
  <div
    class="css-1jswdmc"
  />
</div>
`;
